README.rst

```diff
@@ -26,3 +26,371 @@ Run `tox` to run the whole test suite:
+   ctrl = RIOTCtrl(env=env, application_directory='.')
+   # flash the application
+   ctrl.make_run(['flash'])
+   # run the terminal through a contextmanager
+   with ctrl.run_term():
+       ctrl.term.expect('>')       # wait for shell to start
+       ctrl.term.sendline("help")  # send the help command
+       ctrl.term.expect('>')       # wait for the command result to finnish
+       print(ctrl.term.before)     # print the command result
+   # run without a contextmanager
+   ctrl.start_term()               # start a serial terminal
+   ctrl.term.sendline("help")      # send the help command
+   ctrl.term.expect('>')           # wait for the command result to finnish
+   print(ctrl.term.before)         # print the command result
+   ctrl.stop_term()                # close the terminal
+
+Creating a RIOTCtrl object is done via environments. If empty then all
+configuration will come from the target application makefile. But any
+Make environment variable can be overridden, for example setting
+``BOARD`` to a target ``BOARD`` which is not the default for that
+application.
+
+Any make target used on RIOT devices can be used on the abstraction
+like: ``make flash`` => ``ctrl.make_run(['flash'])``.
+
+``ctrl.start_term()`` (``make term``\ ’s alter ego) by default spawns a
+pexpect child application. From there interactions with the application
+under use can be atomized. In the example below the output of the
+``"help"`` command is captured:
+
+ShellInteractions
+~~~~~~~~~~~~~~~~~
+
+RIOTCtrl provides a minimal extensions by using pexpect replwrap: “[A]
+Generic wrapper for read-eval-print-loops, a.k.a. interactive shells”.
+This implements the perfect wrapper for RIOT shell commands since it
+will wait for a command to finish before returning its output.
+
+RIOT already provides a ``ShellInteraction`` for the ``"help"`` command,
+it can be imported as ``from riotctrl_shell.sys import Help``.
+
+The previous example can be re-written using ``ShellInteraction``:
+
+.. code:: python
+
+   from riotctrl.ctrl import RIOTCtrl
+   from riotctrl.shell import ShellInteraction
+
+   env = {'BOARD': 'native'}
+   # if not running from the application directory the a path must be provided
+   ctrl = RIOTCtrl(env=env, application_directory='.')
+   # flash the application
+   ctrl.flash()                     # alias for ctrl.make_run(['flash'])
+   # shell interaction instance
+   shell = ShellInteraction(ctrl)
+   shell.start_term()               # start a serial terminal
+   print(shell.cmd("help"))         # print the command result
+   shell.stop_term()                # close the terminal
+
+or using the already provided ``Help`` ``ShellInteraction``:
+
+.. code:: python
+
+   class Help(ShellInteraction):
+       """Help ShellInteraction"""
+       @ShellInteraction.check_term
+       def help(self, timeout=-1, async_=False):
+           """Sends the reboot command via the terminal"""
+           return self.cmd("help", timeout, async_)
```

Comment:
That the `Help` shell interaction is printed here, confused me at first. Is this required to be copy-pasted in the code. No => maybe remove it.

Reply:
Linked instead